Alert silences management

alexandre-allard opened the pull request:
**Component**: salt

**Context**: See #2193

**Summary**:
Add a Salt module to manage alert silences.

**Acceptance criteria**:
After having deployed a bootstrap node, being able to put a silence on an alert, list it and then remove it, from the Salt master:
```
salt-run salt.cmd metalk8s_monitoring.get_alerts
salt-run salt.cmd metalk8s_monitoring.add_silence KubeMemOvercommit
salt-run salt.cmd metalk8s_monitoring.get_silences active
salt-run salt.cmd metalk8s_monitoring.delete_silence ${silence_id}
```

---

Closes: #2464

alexandre-allard committed:
salt: add new Salt module for monitoring

for now, this module only manages silences (add,
delete, list) through call to Alertmanager API.
It's also possible to list alerts.

Refs: #2464

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

Ebaneck commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+
+    try:
+        response = requests.request(method, url, verify=False, **kwargs)
+    except Exception as exc:
+        raise CommandExecutionError(
+            "Unable to query Alertmanager API on {}: {}".format(url, str(exc))
+        )
```
I hear it is good practice to use `{!s}` instead of converting with str(exc)

alexandre-allard replied:
I don't know for this, maybe... For me it does the exact same thing, but it's shorter.
Will change it.

gdemonet replied:
Modifiers like this are especially useful when you are passing a complex object to `str.format`:

```python
class MagicSpell:
    incantation = "Wingardium Leviosa"
    effects = ["Levitation", "CharismaBonus"]
    level = 3

    def __repr__(self):
        return (
            "<MagicSpell "
            "incantation={self.incantation} "
            "effect={self.effects!r} "
            "level={self.level!s}>"
        ).format(self=self)
```

It's even more common when using Python 3.6 f-strings:

```python
try:
    # break things
except Exception as exc:
    raise MyError(f"Couldn't break things: {exc!s}")
```

alexandre-allard replied:
done

Ebaneck commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+    return __virtualname__
+
+
+def _requests_alertmanager_api(route, method='GET', **kwargs):
```
I think this is not the most important thing, to begin with, per say! the helper section could come last?
Btw, what is the pep8 standard for this?

alexandre-allard replied:
I don't know for the PEP8 standard, is there anything about that ? I'm not sure. But we don't follow any standard for sure in our modules, should we ?

Ebaneck replied:
None that I know of.

gdemonet replied:
Not sure if PEP8 says anything about this, as it's not really about formatting the code but more about organizing modules. If you want to make sure, feel free to read it and tell us :wink:

On the other hand, we could define our own standards for how to organize these modules, and indeed, where to put helper functions.
As a common practice, I usually try to define pieces of code before they are used. However, in Salt modules, it would probably make sense to have all the public methods first.

slaperche-scality replied:
FWIW, when the language allows it, I usually organize the code is a top-down fashion (i.e the helpers tends to be further down or at the end of the file).

Rationale: I assume that you start reading a file from the top (to read the module doc for instance). Then you read down from here.
If you're only looking for high level info, you'll get what you want in the first functions.
If you need more details, you keep reading further into the file, to dive into the implementation details.

When looking for high level info, you don't have to read each function to know if it's an internal helper (and you need to skip it) or if it's an "interesting" function.


> On the other hand, we could define our own standards for how to organize these modules, and indeed, where to put helper functions.

This could go in our Python best practice, if we come up with something that please everyone.

alexandre-allard replied:
I've moved the helper method to the bottom of the file.

Ebaneck commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+        host, CONTROL_PLANE_INGRESS_PORT, route
+    )
+
+    try:
+        response = requests.request(method, url, verify=False, **kwargs)
+    except Exception as exc:
+        raise CommandExecutionError(
+            "Unable to query Alertmanager API on {}: {}".format(url, str(exc))
+        )
```
This is a request query that might fail for many different reasons, maybe we should handle retries? See example here: https://github.com/scality/metalk8s/pull/2249/files

alexandre-allard replied:
Good point.

alexandre-allard replied:
Done

Ebaneck commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+        'GET',
+    )
+
+    if state is not None:
+        silences = [
+            silence for silence in response
+            if silence['status']['state'] == state
+        ]
+    else:
+        silences = response
```
Could this be a one-liner ??
`silences = [silence for silence in response if silence['status']['state'] == state] if state else response`

alexandre-allard replied:
It could be, but I don't see the point in doing this, other than making the code more compact and less readable.

Ebaneck replied:
Ok, let's keep it as you have done then...

gdemonet replied:
From the Zen of Python:

> Simple is better than complex.

,

> Sparse is better than dense.

and:

> Readability counts.

Ebaneck replied:
Well noted!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

gdemonet commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+        error_msg = 'Missing dependencies: {}'.format(', '.join(MISSING_DEPS))
+        return False, error_msg
+
+    requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
```
I'd rather make sure we trust the CA than go out of our way to ignore warnings globally (what about other consumers of `requests`?). It's the Ingress CA, so you can get the bundle at `/etc/metalk8s/pki/nginx-ingress/ca.[crt|key]`.

gdemonet replied:
Well, that's only if we use the Ingress. But we're inside the cluster, so we shouldn't.

alexandre-allard replied:
Removed

gdemonet commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+            )
+        else:
+            error = (
+                "Malformatted response returned from Alertmanager API: {}: {}"
```
Mmh, I think it'd be either `Misformatted` or `Malformed`

alexandre-allard replied:
Changed to `Malformed`

gdemonet commented on salt/_modules/metalk8s_monitoring.py:
```diff
@@ -0,0 +1,173 @@
+def add_silence(value, name='alertname', is_regex=False, starts_at=None,
+                duration=3600, ends_at=None, time_format='%Y-%m-%dT%H:%M:%S',
+                author='', comment=''):
+    """Add a new silence in Alertmanager.
```
It's an execution module, I think Salt standard practice recommends adding a few examples to use the module from CLI.

gdemonet commented on salt/metalk8s/addons/prometheus-operator/deployed/ingress.sls:
```diff
@@ -0,0 +1,25 @@
+kind: Ingress
+metadata:
+  name: prometheus-operator-alertmanager
+  namespace: metalk8s-monitoring
```
Now that I'm thinking about it, do we **really** need the Ingress to talk to Alertmanager from Salt Master?

It's the first time that we need access to a Service from an execution module, so we didn't really have precedence, but we do look for Service Endpoints in the external pillar `metalk8s_endpoints`, and use those values in some formulas.
IMO, we should do the same here, and maybe refactor slightly to have:
- an execution module for gathering Endpoints for a Service
- this module used in the pillar module
- and use it instead of the Ingress in your `metalk8s_monitoring` module

IIRC, `requests` expose a way to pass multiple hosts to try from (but IIRC, an Endpoint is created if the Pod is Ready, so you shouldn't need it).

alexandre-allard replied:
Removed, I now use alertmanager service endpoint.

gdemonet reviewed (changes requested):
Ingress (even control-plane) objects are for exposing services to the outside. From Salt, we shouldn't need it.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

* @gdemonet

alexandre-allard committed:
salt: add a method to fetch a service endpoints

retrieve the endpoint for a given service, returning
the first IP address and a list of ports.

this method will then be used to retrieve alertmanager
endpoint and communincate with its API to manage
alerts and silences.

Refs: #2464

alexandre-allard committed:
salt: use the common method to fetch service endpoints

this method has been moved to a common place, so that
other modules can use it.

Refs: #2464

alexandre-allard committed:
salt: add metalk8s utils module

add an helper method to automatically retry
http requests.

Refs: #2464

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @ebaneck

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

alexandre-allard commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
